graph.hpp

```diff
@@ -0,0 +1,97 @@
+#ifndef GRAPH_H_INCLUDE
+#define GRAPH_H_INCLUDE
+#include<vector>
+#include<map>
+#include<set>
+#include<memory>
+
+typedef long long int64;
+
+class Graph {
+
+  // Nested class representing node in the graph to manage connections
+  class Node {
+    // Since this class is not used anywhere else, OK to make it a friend
+    friend class Graph;
+    // These represent incoming and outgoing edges.
+    // Reasons for having both sets:
+    // 1. When a node is deleted, this makes it finding out which edges to
+    //    delete much easier
+    // 2. This makes reverse a lot easier
+    std::shared_ptr<std::set<int64> > outgoing;
+    std::shared_ptr<std::set<int64> > incoming;
+    // A bunch of mutator methods to add/delete incoming/outgoing edges from
+    // node
+    void insertOutgoing(int64 to);
+    void insertIncoming(int64 from);
+    void eraseOutgoing(int64 to);
+    void eraseIncoming(int64 from);
+    // Two accessor methods. Return true if there exists an incoming/outgoing
+    // edge
+    bool containsEdgeFrom(int64 from);
+    bool containsEdgeTo(int64 to);
+    Node DeepCopy();
+    public:
+    Node() :
+      outgoing(new std::set<int64>()),
```

Comment:
Ah you bastard-- I see why you used pointers now: I think it's because Reverse() is much faster when you use a pointer instead of actually storing the lists on the Node. Also I see it's not a memory leak because you're using shared_ptr instead of just a normal pointer. Good job.

But yo do me a favor: At Google we like to use unique_ptr instead of shared_ptr because shared_ptr can fuck you really hard if you accidentally leave a dangling reference to it. I actually almost never see shared_ptr in the code for that reason. So can you change the outgoing_ and incoming_ to unique_ptr instead of shared_ptr above? Then in Reverse you can use std::unique_ptr::swap

http://en.cppreference.com/w/cpp/memory/unique_ptr/swap